clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java

```diff
@@ -0,0 +1,145 @@
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Pluggable interface for selecting a preferred read replica given the current set of replicas for a partition
+ * and metadata from the client.
+ */
+public interface ReplicaSelector extends Configurable, Closeable {
+
+    /**
+     * Select the preferred replica a client should use for fetching. If no replica is available, this will return an
+     * empty optional.
+     */
+    Optional<ReplicaView> select(TopicPartition topicPartition,
+                                 ClientMetadata clientMetadata,
+                                 PartitionView partitionView);
+    @Override
+    default void close() throws IOException {
+        // No-op by default
+    }
+
+    @Override
+    default void configure(Map<String, ?> configs) {
+        // No-op by default
+    }
+
+    /**
+     * Holder for all the client metadata required to determine a preferred replica.
+     */
+    interface ClientMetadata {
+        String rackId();
+
+        String clientId();
+
+        InetAddress clientAddress();
+
+        KafkaPrincipal principal();
+
+        String listenerName();
+
+        ClientMetadata NO_METADATA = new ClientMetadata() {
+            @Override
+            public String rackId() {
+                return "";
+            }
+
+            @Override
+            public String clientId() {
+                return "";
+            }
+
+            @Override
+            public InetAddress clientAddress() {
+                return null;
+            }
+
+            @Override
+            public KafkaPrincipal principal() {
+                return null;
+            }
+
+            @Override
+            public String listenerName() {
+                return "";
+            }
+        };
+    }
+
+    /**
+     * View of a partition used by {@link ReplicaSelector} to determine a preferred replica.
+     */
+    interface PartitionView {
+        Set<ReplicaView> replicas();
+
+        /**
+         * Helper method to find the leader among the replicas for this partition
+         */
+        default Optional<ReplicaView> findLeader() {
+            return replicas().stream().filter(ReplicaView::isLeader).findFirst();
+        }
+    }
+
+    /**
+     * View of a replica used by {@link ReplicaSelector} to determine a preferred replica.
+     */
+    interface ReplicaView {
+        /**
+         * Is this replica the leader of its partition
+         */
+        boolean isLeader();
+
+        /**
+         * The endpoint information for this replica (hostname, port, rack, etc)
+         */
+        Node endpoint();
+
+        /**
+         * The log end offset for this replica
+         */
+        long logEndOffset();
+
+        /**
+         * The number of milliseconds (if any) since the last time this replica was caught up to the high watermark.
+         */
+        Optional<Long> lastCaughtUpTimeMs();
```

Comment:
Hmm, what would that be exactly? The fetch time? Or would it remain at zero for the leader replica?

Reply:
Yes, that is what I was wondering 😉 . Possibly this should be a delta since the last caught up time or something. Then for the leader, it would always be 0.

Reply:
Looks like this value only gets updated for follower replicas, so the leader will have the initial value of `0L`. This maps to `Optional.empty()` for a ReplicaView. Sound reasonable?

Reply:
Did you not like my suggestion to treat this as a delta since the last time we were caught up? In that case, the value is just always 0 for a leader.